File: Niv1984/app/src/main/java/com/aaronicsubstances/niv1984/etc/FirebaseFacade.java
package com.aaronicsubstances.niv1984.etc;

import com.aaronicsubstances.niv1984.BuildConfig;
import com.google.firebase.firestore.FirebaseFirestore;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class FirebaseFacade {
    private static final Logger LOGGER = LoggerFactory.getLogger(FirebaseFacade.class);

    public static void getConfItems(VersionCheckCallback cb) {
        Utils.EXECUTOR_INSTANCE.execute(() -> {
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            db.collection("conf-" + BuildConfig.BUILD_TYPE)
                    .limit(1)
                    .get()
                    .addOnSuccessListener(result -> {
                        try {
                            VersionCheckResponse configValue = result.getDocuments().get(0).toObject(VersionCheckResponse.class);
                            LOGGER.debug("Successfully parsed conf document: {}", configValue);
                            Utils.HANDLER_INSTANCE.post(() -> cb.accept(configValue));
                        }
                        catch (Exception ex) {
                            LOGGER.warn("Error interpreting conf document.", ex);
                            Utils.HANDLER_INSTANCE.post(() -> cb.accept(null));
                        }
                    })
                    .addOnFailureListener( exception -> {
                        LOGGER.warn("Error getting conf document.", exception);
                        Utils.HANDLER_INSTANCE.post(() -> cb.accept(null));
                    });
        });
    }
}

File: Niv1984-Plus/app/src/main/java/com/aaronicsubstances/niv1984/data/FirebaseFacade.kt
package com.aaronicsubstances.niv1984.data

import com.aaronicsubstances.niv1984.BuildConfig
import com.aaronicsubstances.niv1984.models.LatestVersionCheckResult
import com.aaronicsubstances.niv1984.ui.foreword.ForewordResource
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import org.slf4j.LoggerFactory
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

object FirebaseFacade {
    private val logger = LoggerFactory.getLogger(javaClass)

    suspend fun fetchAboutResource(relativePath: String): ForewordResource? =
        suspendCoroutine { cont ->
            val db = Firebase.firestore
            db.collection("about-${BuildConfig.BUILD_TYPE}")
                .document(relativePath)
                .get()
                .addOnSuccessListener {
                    val result = it.toObject(ForewordResource::class.java)
                    cont.resume(result)
                }
                .addOnFailureListener {ex ->
                    logger.warn("Error getting about resource at $relativePath: ", ex)
                    cont.resume(null)
                }
        }

    suspend fun getConfItems(): LatestVersionCheckResult? =
        suspendCoroutine { cont ->
            val db = Firebase.firestore
            db.collection("conf-${BuildConfig.BUILD_TYPE}")
                .limit(1)
                .get()
                .addOnSuccessListener { result ->
                    try {
                        val configValue = result.documents[0].toObject(LatestVersionCheckResult::class.java)
                        logger.info("Successfully parsed conf document: $configValue")
                        cont.resume(configValue)
                    }
                    catch (ex: Exception) {
                        logger.warn("Error interpreting conf document.", ex)
                        cont.resume(null)
                    }
                }
                .addOnFailureListener { exception ->
                    logger.warn("Error getting conf document.", exception)
                    cont.resume(null)
                }
        }
}